src/promptflow/tests/executor/unittests/contracts/test_flow.py

```diff
@@ -58,3 +74,536 @@ def test_node_condition_conflict(self):
+        tool2 = Tool(name="tool2", type=ToolType.PYTHON, inputs={}, module="module")
+        node1 = Node(name="node1", tool="tool1", inputs={}, module="yaml")
+        node2 = Node(name="node2", tool="tool2", inputs={}, module="module")
+        tools = [tool1, tool2]
+        nodes = [node1, node2]
+
+        with pytest.raises(FailedToImportModule) as e:
+            Flow._import_requisites(tools, nodes)
+
+        assert str(e.value).startswith("Failed to import modules with error:")
+
+    def test_apply_default_node_variants(self):
+        node_variant = NodeVariant(
+            node=Node(name="print_val_variant", tool=None, inputs={"input2": None}, use_variants=False),
+            description=None,
+        )
+        node_variants = {
+            "print_val": NodeVariants(
+                default_variant_id="variant1",
+                variants={"variant1": node_variant},
+            )
+        }
+        flow1 = Flow(
+            id="test_flow_id",
+            name="test_flow",
+            nodes=[Node(name="print_val", tool=None, inputs={"input1": None}, use_variants=True)],
+            inputs={},
+            outputs={},
+            tools=[],
+            node_variants=node_variants,
+        )
+        # test when node.use_variants is True
+        flow1._apply_default_node_variants()
+        assert flow1.nodes[0].use_variants is False
+        assert flow1.nodes[0].inputs.keys() == {"input2"}
+        assert flow1.nodes[0].name == "print_val"
+
+        flow2 = Flow(
+            id="test_flow_id",
+            name="test_flow",
+            nodes=[Node(name="print_val", tool=None, inputs={"input1": None}, use_variants=False)],
+            inputs={},
+            outputs={},
+            tools=[],
+            node_variants=node_variants,
+        )
+        # test when node.use_variants is False
+        tmp_nodes = flow2.nodes
+        flow2._apply_default_node_variants()
+        assert flow2.nodes == tmp_nodes
+
+    def test_apply_default_node_variant(self):
+        node = Node(name="print_val", tool=None, inputs={"input1": None}, use_variants=True)
+        node_variant = NodeVariant(
+            node=Node(name="print_val_variant", tool=None, inputs={"input2": None}, use_variants=False),
+            description=None,
+        )
+        # test when node_variants is None
+        assert Flow._apply_default_node_variant(node, None) == node
+
+        # test when node.name is not in node_variants
+        node_variants1 = {"test": NodeVariants(default_variant_id="variant1", variants={"variant1": node_variant})}
+        assert Flow._apply_default_node_variant(node, node_variants1) == node
+        node_variants2 = {"print_val": NodeVariants(default_variant_id="test", variants={"variant1": node_variant})}
+        # test when default_variant_id is not in variants
+        assert Flow._apply_default_node_variant(node, node_variants2) == node
```

Comment:
Can we do these tests with ` @pytest.mark.parametrize`?

Reply:
Added parametrize but the pre-commit makes it take up a lot of lines.

Reply:
It's fine. I like the new style.